plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataUpdater.java

```diff
@@ -0,0 +1,237 @@
+import org.sonatype.nexus.repository.maven.internal.maven2.metadata.Maven2Metadata.Snapshot;
+import org.sonatype.nexus.repository.util.TypeTokens;
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.repository.view.payloads.BytesPayload;
+import org.sonatype.nexus.repository.view.payloads.StringPayload;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.hash.HashCode;
+import org.apache.maven.artifact.repository.metadata.Metadata;
+import org.apache.maven.artifact.repository.metadata.SnapshotVersion;
+import org.apache.maven.artifact.repository.metadata.Versioning;
+import org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Reader;
+import org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Writer;
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Maven 2 repository metadata updater.
+ *
+ * @since 3.0
+ */
+public class MetadataUpdater
+    extends ComponentSupport
+{
+  private final MavenFacet mavenFacet;
+
+  private final Maven2MetadataMerger metadataMerger;
+
+  private final MetadataXpp3Reader metadataReader;
+
+  private final MetadataXpp3Writer metadataWriter;
+
+  private final DateFormat dotlessTimestampFormat;
+
+  private final DateFormat dottedTimestampFormat;
+
+  @Inject
+  public MetadataUpdater(final MavenFacet mavenFacet) {
+    this.mavenFacet = checkNotNull(mavenFacet);
+    this.metadataMerger = new Maven2MetadataMerger();
+    this.metadataReader = new MetadataXpp3Reader();
+    this.metadataWriter = new MetadataXpp3Writer();
+    this.dotlessTimestampFormat = new SimpleDateFormat("yyyyMMddHHmmss");
+    this.dotlessTimestampFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+    this.dottedTimestampFormat = new SimpleDateFormat("yyyyMMdd.HHmmss");
+    this.dottedTimestampFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+  }
+
+  /**
+   * Writes/updates metadata, merges existing one, if any.
+   */
+  public void update(final MavenPath mavenPath, final Maven2Metadata metadata) {
+    checkNotNull(mavenPath);
+    checkNotNull(metadata);
+    try {
+      final Metadata oldMetadata = read(mavenPath);
+      if (oldMetadata == null) {
+        // old does not exists, just write it
+        write(mavenPath, toMetadata(metadata));
+      }
+      else {
+        // TODO: compare? unsure is it worth it, as compare would also eat CPU maybe even more that writing would
+        // update old by merging them and write out
+        final Metadata updated = metadataMerger.merge(
+            ImmutableList.of(
+                new MetadataEnvelope("old:" + mavenPath.getPath(), oldMetadata),
```

Comment:
not sure if we actually need it, but might be better to flag MetadataEnvelope as old/new rather than embedding it in the path name?